### PR TITLE
Support ttrpc

### PR DESCRIPTION
**Pull request by Xynnn007**
This PR brings in ttrpc support. Together, `keywrap-grpc`, `keywrap-ttrpc` and `keywrap-native` are now exclusive features. 

Close https://github.com/confidential-containers/image-rs/issues/107

### CODE REVIEW

**Comment by Xynnn007** [comment]
cc @jiangliu @arronwy

**Comment by Xynnn007** [comment]
Hi, @jiangliu @jialez0 @arronwy  this PR is ready for review. 

Hi @HaokunX-intel  please give some suggestions for the new feature `native-aa`, which might influence enclave-cc

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -41,21 +41,21 @@ jobs:
         run: |
           sudo -E PATH=$PATH -s cargo test --all --features default
 
-      - name: Run cargo test - no-default-features
+      - name: Run cargo test - no-default-features,grpc
```
Comment (jiangliu) [inline .github/workflows/ci.yml:44]: This group of tests are trying to test each features individually, so can we avoid add feature grpc to each test case here?

Comment (Xynnn007) [reply .github/workflows/ci.yml:44]: As the features `grpc`, `ttrpc` and `native-aa` are exclusive, we should specify one

**Comment by HaokunX-intel** [comment]
We compile the `enclave agent` with `image-rs = { git = "https://github.com/Xynnn007/image-rs", features = ["occlum_feature", "cosign", "native-aa"], default-features = false, rev = "14885e0134d60277306031a07a8c2468dadb4900" }`, but it complains that
```
error: future cannot be sent between threads safely
   --> src/services/images.rs:81:50
    |
81  |       ) -> ttrpc::Result<image::PullImageResponse> {
    |  __________________________________________________^
82  | |         match self.pull_image(&req).await {
83  | |             Ok(r) => {
84  | |                 println!("Pull image {:?} successfully", r);
...   |
92  | |         }
93  | |     }
    | |_____^ future created by async block is not `Send`
    |
    = help: the trait `Send` is not implemented for `(dyn image_rs::secure_channel::Client + 'static)`
note: future is not `Send` as this value is used across an await
   --> /home/cre/.cargo/git/checkouts/image-rs-65db136907749299/14885e0/src/image.rs:222:75
    |
218 |                 let secure_channel = secure_channel
    |                     -------------- has type `Arc<tokio::sync::Mutex<SecureChannel>>` which is not `Send`
...
222 |                 match credential_for_reference(&reference, secure_channel).await {
    |                                                                           ^^^^^^ await occurs here, with `secure_channel` maybe used later
...
232 |             }
    |             - `secure_channel` is later dropped here
    = note: required for the cast from `impl Future<Output = Result<PullImageResponse, ttrpc::Error>>` to the object type `dyn Future<Output = Result<PullImageResponse, ttrpc::Error>> + Send`

error: future cannot be sent between threads safely
   --> src/services/images.rs:81:50
    |
81  |       ) -> ttrpc::Result<image::PullImageResponse> {
    |  __________________________________________________^
82  | |         match self.pull_image(&req).await {
83  | |             Ok(r) => {
84  | |                 println!("Pull image {:?} successfully", r);
...   |
92  | |         }
93  | |     }
    | |_____^ future created by async block is not `Send`
    |
    = help: the trait `Send` is not implemented for `dyn AsyncRead + Unpin`
note: future is not `Send` as this value is used across an await
   --> /home/cre/.cargo/git/checkouts/image-rs-65db136907749299/14885e0/src/pull.rs:331:63
    |
330 |         let async_decoder = decoder.async_decompress(input_reader);
    |             ------------- has type `Box<dyn AsyncRead + Unpin>` which is not `Send`
331 |         stream_processing(async_decoder, diff_id, destination).await
    |                                                               ^^^^^^ await occurs here, with `async_decoder` maybe used later
332 |     }
    |     - `async_decoder` is later dropped here
    = note: required for the cast from `impl Future<Output = Result<PullImageResponse, ttrpc::Error>>` to the object type `dyn Future<Output = Result<PullImageResponse, ttrpc::Error>> + Send`

error: could not compile `enclave-agent` due to 2 previous errors

```

**Comment by Xynnn007** [comment]
Let's wait for the https://github.com/confidential-containers/ocicrypt-rs/pull/45 to be merged.

**Comment by arronwy** [comment]
> Let's wait for the [confidential-containers/ocicrypt-rs#45](https://github.com/confidential-containers/ocicrypt-rs/pull/45) to be merged.

Merged now :-)

**Comment by Xynnn007** [comment]
Hi, @HaokunX-intel I've fixed the error you've mentioned. Could you please try again? The dependency item is
```toml
image-rs = { git = "https://github.com/Xynnn007/image-rs", features = ["enclave-cc", "signature-cosign"], default-features = false, rev = "5f6c5b5" }
```

File: Cargo.toml [hunk 2]
```diff
@@ -25,33 +25,41 @@ log = "0.4.14"
...
 serde_yaml = { version = "0.8", optional = true }
 sha2 = ">=0.10"
+
 # Latest version of sigstore conflicts with sequoia-openpgp, so stick to v0.3
 sigstore = { version = "0.3.0",  optional = true}
 strum = { version = "0.24", features = ["derive"] }
 strum_macros = "0.24"
 tar = "0.4.37"
-tokio = "1"
+tokio = {version = "1.0", features = ["full"]}
```
Comment (jiangliu) [inline Cargo.toml:42]: Please help to explicitly specify features needed:)

Comment (Xynnn007) [reply Cargo.toml:42]: Nice suggestion, I'm going to

**Review by jiangliu — APPROVED** [review]

File: Cargo.toml [hunk 3]
```diff
@@ -25,9 +25,10 @@ log = "0.4.14"
 nix = { version = "0.26", optional = true }
 oci-distribution = "0.9.4"
 oci-spec = "0.5.8"
-ocicrypt-rs = { git = "https://github.com/confidential-containers/ocicrypt-rs", rev = "52c9a3c", default-features = false, features = ["keywrap-jwe", "async-io"], optional = true }
+ocicrypt-rs = { git = "https://github.com/confidential-containers/ocicrypt-rs", rev = "453c699", default-features = false, features = ["keywrap-jwe", "keywrap-keyprovider-cmd", "keywrap-keyprovider-grpc", "keywrap-keyprovider-native", "keywrap-keyprovider-ttrpc", "async-io"], optional = true }
```
Comment (arronwy) [inline Cargo.toml:28]: Default `ocicrypt-rs` don't need add  `keywrap-keyprovider-*` features

File: Cargo.toml [hunk 4]
```diff
@@ -63,15 +68,16 @@ members = ["libs/test-utils"]
...
 kata-cc = ["encryption", "keywrap-grpc", "snapshot-overlayfs", "signature-cosign"]
 kata-cc-s390x = ["encryption", "keywrap-grpc", "snapshot-overlayfs", "signature-simple"]
 enclave-cc = ["encryption", "keywrap-native", "eaa-kbc", "snapshot-unionfs", "signature-simple"]
 
 encryption = ["ocicrypt-rs/block-cipher"]
 keywrap-cmd = ["ocicrypt-rs/keywrap-keyprovider-cmd"]
-keywrap-grpc = ["ocicrypt-rs/keywrap-keyprovider-grpc", "prost", "tonic"]
+keywrap-grpc = ["ocicrypt-rs/keywrap-keyprovider-grpc", "prost", "tonic", "tonic-build"]
 keywrap-native = ["ocicrypt-rs/keywrap-keyprovider-native", "attestation_agent"]
+keywrap-ttrpc = ["dep:ttrpc", "dep:protobuf", "ttrpc-codegen"]
```
Comment (arronwy) [inline Cargo.toml:80]: `ocicryp-rs/keywrap-keyprovider-ttrpc` feature can add here

**Review by arronwy — APPROVED** [review]
Thanks @Xynnn007 , LGTM! Only some minor suggestions. BTW @HaokunX-intel can you double check whether this version fixed the error you metioned during build with `enclave-agent`

File: Cargo.toml (hunk 4 above)
Comment (Xynnn007) [reply Cargo.toml:80]: Thanks! Nice suggestions! I'd force push after @HaokunX-intel had ensured the fixes.

**Pull request merged**